### PR TITLE
Homework 12

### PR DESCRIPTION
**Pull request by jurisk**

### CODE REVIEW

File: homeworks/12-mobx-mars/src/state/state.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    load: () => void
+
+    selectedPhotos(): readonly PhotoPair[]
+
+    favouritePhotos(): readonly PhotoPair[]
+    addFavourite: (photoId: PhotoId) => void
+    removeFavourite: (photoId: PhotoId) => void
+}
+
+class StateImpl {
```
Comment (jurisk) [inline homeworks/12-mobx-mars/src/state/state.ts:22]: I was considering to split it up somehow to make it less "monolith", but didn't think of the best way to do it, so did not do it yet.

File: homeworks/12-mobx-mars/src/features/photos/photos.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                />)}
+            </div>
+        </div>
+    ) : <div>No photos.</div>
+}
+
+export const PhotoList = observer(() =>
+    <ShowPhotos
+        photoPairs={state.selectedPhotos()}
+        addFavourite={(x) => state.addFavourite(x)}
```
Comment (jurisk) [inline homeworks/12-mobx-mars/src/features/photos/photos.tsx:64]:
```typescript
addFavourite={state.addFavourite}
``` 

did not work here!

**Comment by jurisk** [comment]
Known limitation - "default CSS" used (as in the previous app).

File: homeworks/12-mobx-mars/src/state/state.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        this.selectedSol = sol
+    }
+
+    load(): void {
+        loadRoverPhotos(DefaultRover, this.selectedSol).then((results) =>
+            this.addReceivedPhotos(this.selectedSol, results)
+        )
+    }
+
+    addReceivedPhotos(sol: Sol, results: readonly Photo[]) {
```
Comment (avilde) [inline homeworks/12-mobx-mars/src/state/state.ts:46]: parameter `sol` is not used

File: homeworks/12-mobx-mars/src/api.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    img_src: string;
+    rover: {
+        name: string;
+    }
+    camera: {
+        name: string;
+    }
+}
+
+function loadSinglePageOfRoverPhotos(rover: Rover, sol: Sol, page: PageNumber = 1 as PageNumber): Promise<readonly Photo[]> {
```
Comment (avilde) [inline homeworks/12-mobx-mars/src/api.ts:21]: The API endpoint used here returns `{ photos: PhotoJson[]}` object.
I would simplify this in the following way:

```typescript
interface JsonResponse {
    photos: PhotoJson[]
}

function convertPhotoData(response: JsonResponse) {
    return response.photos.map((photoJson) =>
        ({
            id: photoJson.id as PhotoId,
            imageUrl: photoJson.img_src as Url,
            description: `${photoJson.id} - ${photoJson.rover.name} - ${photoJson.camera.name}`,
        })
    )
}

function loadSinglePageOfRoverPhotos(rover: Rover, sol: Sol, page: PageNumber = 1 as PageNumber): Promise<readonly Photo[]> {
    return fetch(`https://api.nasa.gov/mars-photos/api/v1/rovers/${rover}/photos?sol=${sol}&api_key=${ApiKey}&page=${page}`)
        .then(response => response.json() as Promise<JsonResponse>)
        .then(convertPhotoData)
}
```

Also error handling should be added if `fetch` failed to retrieve data.

**Review by avilde — COMMENTED** [review]
# Review

## Good parts
* Mobx version of application is working fine 👍🏻 

## What can be improved
* error handling for fetch
* some CSS wouldn't heart for the buttons

## Summary
Application has been successfully converted from use of Redux to Mobx. Good job!

File: homeworks/12-mobx-mars/src/api.ts (hunk 4 above)
Comment (jurisk) [reply homeworks/12-mobx-mars/src/api.ts:35]: Thanks, this is better!

**Pull request merged**